tests/unit/test_retry.py

```diff
@@ -0,0 +1,184 @@
+from datetime import datetime, timedelta
+from os import getenv
+from typing import AsyncIterator
+from unittest.mock import AsyncMock, MagicMock
+
+import pytest
+
+from jbi.queue import QueueItem, get_dl_queue
+from jbi.retry import retry_failed
+
+RETRY_TIMEOUT_DAYS = getenv("RETRY_TIMEOUT_DAYS", 7)
+
+
+def add_iter(obj):
+    mock = MagicMock()
+    mock.__aiter__.return_value = obj
+    return mock
+
+
+def iter_error():
+    mock = MagicMock()
+    mock.__aiter__.return_value = None
+    mock.__aiter__.side_effect = Exception("Throwing an exception")
+    return mock
+
+
+@pytest.mark.asyncio
+async def test_retry_empty_list(caplog):
+    retrieve = AsyncMock(return_value={})
+    get_dl_queue().retrieve = retrieve
```

Comment:
In other tests we mock the instantiation of the singleton in a fixture:

https://github.com/mozilla/jira-bugzilla-integration/blob/dca677f6ecb94607c3468da85a2477283833e5a3/tests/conftest.py#L110-L118

Maybe we want to follow the same pattern?

I think there is also an opportunity to leverage the composition pattern on the queue backend.
We could chose to not mock the queue methods but inject a mocked backend instead.

(completely untested)
```py
@pytest.fixture()
def mocked_queue(request):
    class MockedBackend(QueueBackend):
         get_all = AsyncMock()
         size = MagicMock()
         ...
    
    with mock.patch("jbi.queue.get_dl_queue") as mocked:
        mocked.return_value.backend = MockedBackend()
        yield
    jbi.queue.get_dl_queue.cache_clear()

```

AFAIU, the backend methods could be have their return value set like this:

```py
async def test_retry_empty_list(caplog, mocked_queue):
     mocked_queue.backend.get_all.return_value=[]
```

Reply:
In addition to these good suggestions, I tried playing around with this and came up with:


```python
@pytest.fixture
def queue():
    return MagicMock(spec=DeadLetterQueue)


@pytest.fixture
def executor():
    return MagicMock(spec=execute_action)


async def aiter_sync(iterable):
    for i in iterable:
        yield i


@pytest.mark.asyncio
async def test_retry_success(caplog, queue, queue_item_factory, executor):
    queue.retrieve.return_value = {1: aiter_sync([queue_item_factory()])}
    metrics = await retry_failed(queue, executor)
    queue.retrieve.assert_called_once()  # no logs should have been generated
    queue.done.assert_called_once()  # item should be marked as complete
    executor.assert_called_once()  # item should have been processed
    assert metrics == {
        "bug_count": 1,
        "events_processed": 1,
        "events_skipped": 0,
        "events_failed": 0,
        "bugs_failed": 0,
    }
```

It's more dependency injection, but it's nice because we're explicitly building an object with behavior we expect for the test, then passing it directly to the function under test